Support parent FakeTensorMode to support immutable/mutable split case

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113770
* #113769
* #113768
* #113767
* __->__ #113653
* #113651

Relevant comment:

```
        # The parent fake tensor mode lets you specify a fake tensor mode that
        # you should fakeify any real tensors through first, before you
        # fakeify them again here.  This is most useful if you have an
        # immutable fake tensor mode you've created to keep track of the
        # initial memoized fakeifications of real tensors; you can then
        # chain fresh fake tensor modes off of the original fake tensor
        # mode to reuse all of the properties you set, while having a
        # completely distinct universe of fake tensors which you can mutate to
        # your hearts content.
```

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Support parent FakeTensorMode to support immutable/mutable split case

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

eellison reviewed (commented):
```
you can then
        # chain fresh fake tensor modes off of the original fake tensor
        # mode to reuse all of the properties you set, while having a
        # completely distinct universe of fake tensors which you can mutate to
        # your hearts content.
```

Maybe just for my own edification - but why is this preferable to storing the previous attributes of a Tensor whenever you are about to run metadata mutation on it ?

ezyang commented:
@eellison the fake tensors need to have accurate aliasing information, and that's pretty difficult to preserve if you have aliasing destroying metadata mutation ops like `set_`

eellison commented:
aliasing is a StorageWeakRef and a storage_offset

voznesenskym committed:
Update on "Support parent FakeTensorMode to support immutable/mutable split case"


Relevant comment:

```
        # The parent fake tensor mode lets you specify a fake tensor mode that
        # you should fakeify any real tensors through first, before you
        # fakeify them again here.  This is most useful if you have an
        # immutable fake tensor mode you've created to keep track of the
        # initial memoized fakeifications of real tensors; you can then
        # chain fresh fake tensor modes off of the original fake tensor
        # mode to reuse all of the properties you set, while having a
        # completely distinct universe of fake tensors which you can mutate to
        # your hearts content.
```

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@eellison my point is, if you're doing "aliasing is a StorageWeakRef and a storage_offset", you might as well just have the entire fake tensor while you're at it lol. Because remember you have to storage ref to *something*...

eellison commented:
I think doing a clone and storing the prior fake tensor is reasonable. The number of metadata mutating ops we acutally run is very very small.

ezyang commented:
What will you do when there is an operation that modifies the storage size

bdhirsh commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -333,6 +335,18 @@ def from_real_tensor(
+                dynamic_dims=dynamic_dims,
+                constraint_dims=constraint_dims,
+                memoized_only=memoized_only,
+            )
```
I guess one (non-obvious) invariant now is that if you create a `FakeTensorMode` with a parent, then its `tensor_memo` cache will always be key'd on `FakeTensor`s instead of real tensors. I guess there's nothing wrong with that? (mostly just calling it out)

ezyang replied:
Yep! In fact, we actually do this already, when you run _dynamo.export with inputs that are already fake tensors (e.g., because you used fake tensors to avoid loading the model for tracing)

bdhirsh commented on test/test_fake_tensor.py:
```diff
@@ -554,6 +554,15 @@ def test_jagged_fake_to_fake_preserved(self):
+    def test_fake_parent(self):
+        shape_env = ShapeEnv()
+        mode1 = FakeTensorMode(shape_env=shape_env)
+        mode2 = FakeTensorMode(parent=mode1, shape_env=shape_env)
```
Should we just add an assert that if you initialize a `FakeTensorMode` with a parent, they must have the same ShapeEnv?

I suppose you could have a parent with a different ShapeEnv (or even without a ShapeEnv), but that seems like it adds a lot of degrees of freedom and I'm not sure what the benefit is.

ezyang replied:
I can add the assert if it makes you happy. To me it does make sense to have non-matching shape env but it is true there is no concrete use case for it

bdhirsh replied:
ok nvm then seems fine to leave

voznesenskym committed:
Update on "Support parent FakeTensorMode to support immutable/mutable split case"


Relevant comment:

```
        # The parent fake tensor mode lets you specify a fake tensor mode that
        # you should fakeify any real tensors through first, before you
        # fakeify them again here.  This is most useful if you have an
        # immutable fake tensor mode you've created to keep track of the
        # initial memoized fakeifications of real tensors; you can then
        # chain fresh fake tensor modes off of the original fake tensor
        # mode to reuse all of the properties you set, while having a
        # completely distinct universe of fake tensors which you can mutate to
        # your hearts content.
```

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Support parent FakeTensorMode to support immutable/mutable split case"


Relevant comment:

```
        # The parent fake tensor mode lets you specify a fake tensor mode that
        # you should fakeify any real tensors through first, before you
        # fakeify them again here.  This is most useful if you have an
        # immutable fake tensor mode you've created to keep track of the
        # initial memoized fakeifications of real tensors; you can then
        # chain fresh fake tensor modes off of the original fake tensor
        # mode to reuse all of the properties you set, while having a
        # completely distinct universe of fake tensors which you can mutate to
        # your hearts content.
```

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Support parent FakeTensorMode to support immutable/mutable split case"


Relevant comment:

```
        # The parent fake tensor mode lets you specify a fake tensor mode that
        # you should fakeify any real tensors through first, before you
        # fakeify them again here.  This is most useful if you have an
        # immutable fake tensor mode you've created to keep track of the
        # initial memoized fakeifications of real tensors; you can then
        # chain fresh fake tensor modes off of the original fake tensor
        # mode to reuse all of the properties you set, while having a
        # completely distinct universe of fake tensors which you can mutate to
        # your hearts content.
```

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Support parent FakeTensorMode to support immutable/mutable split case"


Relevant comment:

```
        # The parent fake tensor mode lets you specify a fake tensor mode that
        # you should fakeify any real tensors through first, before you
        # fakeify them again here.  This is most useful if you have an
        # immutable fake tensor mode you've created to keep track of the
        # initial memoized fakeifications of real tensors; you can then
        # chain fresh fake tensor modes off of the original fake tensor
        # mode to reuse all of the properties you set, while having a
        # completely distinct universe of fake tensors which you can mutate to
        # your hearts content.
```

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]
